doc/developer/design/20230119_ir_normalization.md

```diff
@@ -0,0 +1,147 @@
+For example, the `Union` operators may have two inputs that can be canceled (because one is the `Negate` of the other);
+the transformation that removes cancelable pairs does not need to be part of normalization.
+
+In the future we may want to include some optimizing transformations that are sufficiently easy and obvious. 
+For the moment, we can rely on the fact that we plan to use normalization as part of the optimizer, and all optimizations will be run alongside normalization.
+
+## Description
+
+<!--
+// Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+// If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+I came up with the following components of syntactic normalization.
+These are meant to be structural changes to a `MirRelationExpr`: those that do not risk a change in meaning, by restricting themselves to very relatively simple actions on the shape of the AST.
+Whether a transform is "syntactic" or an action "relatively simple" is up for debate; I hope we'll know these when we see them.
+
+### Simplification
+
+Some operators are just more complicated forms of other operators.
+A `Reduce` with a constant argument can move that argument to a `Map`.
+A `Join` with one input could be a `Filter`.
+A `Join` with a constant argument *may* be able to express that argument as a `Map`. 
+
+Some of these skate close to optimizations.
+Turning an entire `Reduce` into a `Map` relies on the `group_key` being a unique key for the input collection.
+I think we should start judiciously here, especialy as the complexity can spiral.
+Again, we'll be running normalization alongside more sophisticated transforms that can perform more interesting optimizations.
+
+### Movement
+
+Many operators commute.
+When it is appropriate to do so, we should move commuting operators to canonical representations.
+For example, contiguous blocks of `Map`, `Filter`, and `Project` operators can be re-ordered to be `Map`s, `Filter`s, and then `Project`s.
+A tree of `Union` operators may benefit from having linear operators (as above, plus `Negate`, `FlatMap`) pushed down to their inputs, that multiple `Union`s can be merged.
+A tree of `Join` operators may benefit from having linear operators lifted to its root, that multiple `Join`s can be merged.
+
+Group-wise operators like `Reduce`, `Threshold`, `TopK`, etc. (and `ArrangeBy`) may need none of this.
+
+There are *optimizations* that allow us to move e.g. filters through the keys of reductions, but I think there may be some (debatable) value in doing structural and semantic normalization independently.
+Moving e.g. filters great distances may have value, but different transforms may disagree about the direction of movement (up or down) and we would like to keep normalization generally useful.
```

Comment:
> but different transforms may disagree about the direction of movement

I was hoping that we can start with a set of transformations that agree on that direction.

Reply:
I bet the only (initial) agreement will be "no movement that cannot be undone", which is maybe "push down through `Union`, pull up through `Join`, and order within `MFP` in that order. "Hoist all predicates to the root" is for example not a thing that the "physical planning" half of the pipeline wants, but is possibly what the logical optimizer wants.

But seems great if the conclusion is "normalization only gets to do things that everyone wants".